### ISSUE TITLE
How to set the batch dimension with apex.parallel.DistributedDataParallel?

### ISSUE THREAD

**Issue by visionscaper** [NVIDIA/apex#494]
This is more a question than an issue: `apex.parallel.DistributedDataParallel` does not have a `dim` argument, like the original version of the PyTorch `DistributedDataParallel`,  to set the batch dimension. 

In my case, currently my batch dimension is the second dimension in the batch data. Although I can't set the batch dimension for `apex.parallel.DistributedDataParallel` it seems to work. How does this `DistributedDataParallel` implementation know what my batch dimension is? How can I explicitly specify it for distributed data parallel training?

**Comment by mcarilli** [comment]
Pytorch DistributedDataParallel implements a kind of nested DataParallel + DistributedDataParallel combo, if you want it to.  It permits multiple processes with each process handling (potentially) multiple GPUs.  Within each process, the data is chunked (along the batch dimension) across that process's GPUs in a manner similar to DataParallel.  I think the dim argument is only relevant to the intra-process parallelization/chunking (in other words, if you are only using one GPU per process, it doesn't matter).  Apex DistributedDataParallel only supports one GPU per process in any case, so I don't think you need to worry about which dimension is your batch dimension.  Each process will create gradients for its params, and these gradients will be allreduced, which is what you want.

These days `torch.nn.parallel.Distributed.DataParallel` is pretty darn good, and more robust against upstream changes because it's packaged as part of Pytorch core.  Apex DistributedDataParallel mostly exists to support internal use cases, and to avoid breaking codes that currently rely on it.

I recommend using `torch.nn.parallel.DistributedDataParallel` if possible, which can be substituted for Apex DDP simply by changing the DDP constructor call:  https://github.com/NVIDIA/apex/blob/master/examples/simple/distributed/distributed_data_parallel.py#L49-L52.  The example shows using one GPU per process, which is our typical usage regardless of whether we choose apex DDP or torch DDP.

**Comment by visionscaper** [comment]
Ah yes, of course, there is no chunking of batches when having one GPU per process. Thanks a lot for the elaborate answer! Good to know that I should use `torch.nn.parallel.DistributedDataParallel`, will adapt my code now.
